package/src/index/mod.rs

```diff
@@ -330,29 +337,67 @@ impl PackageIndex<Exclusive> {
 /// The identifier of a package in the package index.
 #[derive(Clone, PartialEq, Eq, Debug, Hash, Deserialize, PartialOrd, Ord)]
 pub enum Id {
-    Github { org: String, name: String },
+    Github {
+        org: String,
+        name: String,
+        #[serde(default)]
+        path: RelativePath,
+    },
 }
 
 impl Id {
     /// Returns the path (relative to the package index base directory) where this
     /// package should be stored.
     pub fn path(&self) -> PathBuf {
         match self {
-            Id::Github { org, name } => PathBuf::from(format!("github/{org}/{name}")),
+            Id::Github { org, name, path } if path.is_empty() => {
+                let mut p = PathBuf::from("github");
+                p.push(org);
+                p.push(name);
+                p
+            }
+            // A package that lives in a subdirectory of a git repo gets its
+            // name encoded to be unique. We can't put files for `github:nickel-lang/js2n/lib`
+            // in a subdirectory of the location for `github:nickel-lang/js2n` because their
+            // could be conflicts.
+            //
+            // Instead, we "encode" the name as `github/nickel-lang%@js2n%@lib` so
+            // it won't clash with anything. The special characters are not allowed in
+            // github repo names, so there can't be collisions.
+            Id::Github { org, name, path } => {
+                let mut p = PathBuf::from("github");
+                p.push(org);
+                let mut dir = name.to_owned();
+                for c in path.components() {
+                    dir.push_str("%@");
```

Comment:
Any reason to use `%@` instead of a single character here, such as `@`? To make it clearer that it's mangled?

```suggestion
                    dir.push_str("%@");
```